File: static/script/game.js
var game_setting = null;
var Method = null;
var needready = false;
var images = {};

var player_ready = 0;
var players_num = 0;
var role = null;
var special_power = null;
var hasPercival = false;

var assassin = false;
var chosing = false;
var confirming = false;
var record = null;

var round = 0;
var tokenNeed = null;
var captain = null;
var teamates = [];

function loadImages() {
	$.each(images, function(name, image) {
		if (!(image instanceof Object)) {
			images[name] = new Image();
			images[name].src = image;
		}
	});
}

function ready() {
	if (needready) {
		socket.send(JSON.stringify({
			method: Method.READY,
		}));
		// 
	}
}

function displayCard(players) {
	$.each(players, function(i, player) {
		var div = $("<div>")[0];
		div.classList.add("player-card");
		div.id = "player-" + player[0];
		div.addEventListener('click', clickPlayer);

		var num = $("<div>")[0];
		num.innerHTML = i + 1;
		num.classList.add("number");

		var img = images.unknown.cloneNode();
		img.classList.add("character");
		
		var name = $("<div>")[0];
		name.innerHTML = player[1];
		name.classList.add("name");

		div.appendChild(num);
		div.appendChild(img);
		div.appendChild(name);

		$("#cards")[0].appendChild(div);
	});
}

function displayMyself() {
	var ele = $("#player-" + user_id)[0];
	ele.removeChild($("#player-" + user_id + " .character")[0]);

	var img = null;
	if (role != "SERVANT" && role != "EVIL") {
		img = images["q_" + role.toLowerCase()].cloneNode();
	}
	else if (role == "SERVANT") {
		img = images["q_" + role.toLowerCase() + "_1"].cloneNode();
	}
	else if (role == "EVIL") {
		img = images["q_" + role.toLowerCase() + "_1"].cloneNode();
	}
	img.classList.add("character");
	ele.insertBefore(img, $("#player-" + user_id + " .name")[0]);
}

function clickPlayer(event) {
	if (assassin && role == "ASSASSIN") {
		$.each(event.path, function(_, ele) {
			try {
				if (ele.classList.contains("player-card")) {
					var player_id = ele.id.replace("player-", "");

					// 刺殺梅林
					socket.send(JSON.stringify({
						method: Method.ASSASSIN,
						user_id: player_id,
					}));
				}
			} catch (e) {}
		});
		return;
	}

	if (!chosing || (captain != user_id)) {
		return;
	}

	$.each(event.path, function(_, ele) {
		try {
			if (ele.classList.contains("player-card")) {
				var player_id = ele.id.replace("player-", "");

				if (teamates.includes(player_id)) {
					$("#black-bg").show();
					$("#waiting").show();

					// tell server remove player_id
					socket.send(JSON.stringify({
						method: Method.UNCHOSETEAMATE,
						user_id: player_id,
					}));
				}
				else {
					if (teamates.length >= tokenNeed["numbers"][round]) {
						return;
					}

					$("#black-bg").show();
					$("#waiting").show();

					//  tell server add player_id
					socket.send(JSON.stringify({
						method: Method.CHOSETEAMATE,
						user_id: player_id,
					}));
				}
				
				return false;
			}
		} catch(e) {
			console.log(e)
		}
	});
}

function confirmTeam() {
	if (captain != user_id) {
		return;
	}

	if (teamates.length == tokenNeed["numbers"][round]) {
		socket.send(JSON.stringify({method: Method.COMFIRMTEAM}));
	}
}

function approveTeam() {
	$("#black-bg").show();
	$("#waiting").show();
	socket.send(JSON.stringify({method: Method.APPROVE}));
}

function rejectTeam() {
	$("#black-bg").show();
	$("#waiting").show();
	socket.send(JSON.stringify({method: Method.REJECT}));
}

function missionSuccess() {
	$("#black-bg").show();
	$("#waiting").show();
	socket.send(JSON.stringify({method: Method.SUCCESS}));
}
function missionFail() {
	$("#black-bg").show();
	$("#waiting").show();
	socket.send(JSON.stringify({method: Method.FAIL}));
}

function startHandleMethod() {
	socket.onmessage = function (event) {
		data = JSON.parse(event.data);

		if (data.method == Method.WAITING) {
			players_num = data.players_num;
			$("#waiting-number")[0].innerHTML = players_num;
		}
		else if (data.method == Method.NEEDREADY) {
			$("#ready")[0].classList.remove("disable");
			needready = true;

			// Automatically Ready
			// ready();
		}
		else if (data.method == Method.CONFIRMREADY) {
			if (data.user == user_id) {
				$("#ready")[0].classList.add("active");
			}

			player_ready = data.player_ready;
			$("#ready-number")[0].innerHTML = player_ready;
		}
		else if (data.method == Method.START) {
			needready = false;
			$("#prepare").hide();
			$("#game").show();

			role = data.role;
			special_power = data.special_power;
			hasPercival = data.has_percival;
			players_num = data.players.length;
			tokenNeed = data.token_need;

			$("#board")[0].appendChild(images["b" + players_num]);

			displayCard(data.players);
			displayMyself();
			playeStory();
		}
		else if (data.method == Method.CHOSECAPTAIN) {
			captain = data.captain;
			round = data.round;
			chosing = data.chosing;
			failed = data.failed;
			confirming = false;
			teamates = [];

			// Add captain, teamate, foldedmission vote img to every players
			if ($(".captain").length == 0) {
				$.each($(".player-card"), function(_, card) {
					var captain = images.captain.cloneNode();
					captain.classList.add("captain");

					var teamate = images.mission.cloneNode();
					teamate.classList.add("teamates")

					var vote = images.vote.cloneNode();
					vote.classList.add("vote");

					var foldedmission = images.foldedmission.cloneNode();
					foldedmission.classList.add("foldedmission");

					card.appendChild(captain);
					card.appendChild(teamate);
					card.appendChild(vote);
					card.appendChild(foldedmission);
				});
			}

			$("#failed")[0].innerHTML = "";
			for (var i=0;i<data.failed;i++) {
				$("#failed")[0].appendChild(images.evil_token.cloneNode());
			}

			setTimeout(function() {
				$("#black-bg").hide();
				$("#waiting").hide();
				$("#vote").hide();
				$(".captain").hide();
				$(".teamates").hide();
				$(".vote").hide();
				$(".foldedmission").hide();
				$("#player-" + data.captain + " .captain").show();
			}, 1000);
		}
		else if (data.method == Method.CHOSENTEAMATE) {
			$("#black-bg").hide();
			$("#waiting").hide();
			teamates = data.teamates;

			$.each($(".player-card"), function(i, card) {
				var id = card.id.replace("player-", "")
				if (teamates.includes(id)) {
					$("#player-" + id + " .teamates").show(300);
				}
				else {
					$("#player-" + id + " .teamates").hide(300);
				}
			});

			if (teamates.length == tokenNeed["numbers"][round] && captain == user_id) {
				// 
				$("#confirm").show();
				$("#confirm")[0].style.width = $("#board > img").width() + "px";
				$("#confirm")[0].style.height = $("#board > img").height() + "px";
			}
			else {
				$("#confirm").hide();
			}
		}
		else if (data.method == Method.ASKAPPROVAL) {
			chosing = data.chosing;
			confirming = data.confirming;

			// Make sure user see right teamates
			teamates = data.teamates;

			$.each($(".player-card"), function(i, card) {
				var id = card.id.replace("player-", "")
				if (teamates.includes(id)) {
					$("#player-" + id + " .teamates").show(300);
					$("#player-" + id)[0].style.zIndex = 15;
				}
				else {
					$("#player-" + id + " .teamates").hide(300);
					$("#player-" + id)[0].style.zIndex = 9;
				}
			});

			// 
			if ($("#approve")[0].children.length == 0) {
				$("#approve")[0].appendChild(images.approve);
			}
			if ($("#reject")[0].children.length == 0) {
				$("#reject")[0].appendChild(images.reject);
			}

			$(".vote").hide();
			$("#vote").show();
			$("#confirm").hide();

			// Automatically vote reject
			// approveTeam();
		}
		else if (data.method == Method.VOTECONFIRM) {
			$("#black-bg").hide();
			$("#waiting").hide();

			if (data.voter == user_id) {
				$("#vote").hide(300);
			}

			$("#player-" + data.voter + " .vote").show(300);
		}
		else if (data.method == Method.ASKSUCCESS) {
			if (data.teamate) {
				if ($("#success")[0].children.length == 0) {
					$("#success")[0].appendChild(images.success);
				}
				if ($("#fail")[0].children.length == 0) {
					$("#fail")[0].appendChild(images.fail);
				}

				$("#mission").show();

				if (role == "MERLIN" || role == "PERCIVAL" || role == "SERVANT") {
					$("#fail").hide();
				}
			}

			$("#vote").hide();
			$(".vote").hide();
			$("#black-bg").hide();
			$("#waiting").hide();
		}
		else if (data.method == Method.MISSIONCONFIRM) {
			if (data.voter == user_id) {
				$("#black-bg").hide();
				$("#waiting").hide();
				$("#mission").hide(300);
			}

			$("#player-" + data.voter + " .foldedmission").show(300);
		}
		else if (data.method == Method.GAMERECORD) {
			$.each($(".player-card"), function(i, card) {
				card.style.zIndex = 9;
			});

			// display record under cards
			record = data.record;
			if (data.record[round].resault.good_evil == "good") {
				$("#tokens")[0].appendChild(images.good_token.cloneNode());
			}
			else {
				$("#tokens")[0].appendChild(images.evil_token.cloneNode());
			}

			teamates = [];
			data.round = data.round;
			$(".foldedmission").hide(300);
			$("#mission").hide(300);
		}
		else if (data.method == Method.KILLMERLIN) {
			assassin = true;

			$.each(data.evils_role_map, function(id, role) {
				var ele = $("#player-" + id)[0];
				ele.removeChild($("#player-" + id + " .character")[0]);

				var img = null;
				if (role != "EVIL") {
					img = images["q_" + role.toLowerCase()].cloneNode();
				}
				else if (role == "EVIL") {
					img = images["q_" + role.toLowerCase() + "_1"].cloneNode();
				}
				img.classList.add("character");
				ele.appendChild(img);
			});

			$("#tokens")[0].appendChild(images.good_token.cloneNode());

			if (role == "ASSASSIN") {
				$("#black-bg").show(300);
				$("#assassin-merlin").show(300);

				setTimeout(function() {
					$("#black-bg").hide(50);
					$("#assassin-merlin").hide(50);
				}, STORYTIME);
			}

			setTimeout(function() {
				$("#vote").hide();
				$("#mission").hide();
				$(".captain").hide();
				$(".foldedmission").hide();
				$(".vote").hide();
				$(".teamates").hide();
			}, 1000);
		}
		else if (data.method == Method.END) {
			$.each(data.role_map, function(id, role) {
				var ele = $("#player-" + id)[0];
				ele.removeChild($("#player-" + id + " .character")[0]);

				var img = null;
				if (role != "SERVANT" && role != "EVIL") {
					img = images["q_" + role.toLowerCase()].cloneNode();
				}
				else if (role == "SERVANT") {
					img = images["q_" + role.toLowerCase() + "_1"].cloneNode();
				}
				else if (role == "EVIL") {
					img = images["q_" + role.toLowerCase() + "_1"].cloneNode();
				}
				img.classList.add("character");
				ele.appendChild(img);
			});

			if (data.add_token == 1) {
				$("#tokens")[0].appendChild(images.evil_token.cloneNode());
			}

			setTimeout(function() {
				$("#vote").hide();
				$("#mission").hide();
				$(".captain").hide();
				$(".foldedmission").hide();
				$(".vote").hide();
				$(".teamates").hide();
			}, 1000);
			
			if (data.win == "good") {
				alert("好人獲勝");
			}
			else {
				alert("壞人獲勝");
			}
		}
		else if (data.method == Method.DISCONNECT) {
			players_num = data.players_num;
			player_ready = data.player_ready;
			$("#waiting-number")[0].innerHTML = players_num;
			$("#ready-number")[0].innerHTML = player_ready;
		}
		else {
			console.log(data)
		}
	}
}

$(document).ready(function() {
	$.ajax({
		url: "/game_setting",
		success (msg) {
			game_setting = msg;
			$.each(msg.images, function(_, image) {
				var name = image.substring(0, image.find("."));
				images[name] = "/static/images/" + image;
			});

			Method = {};
			$.each(msg.method, function(_, method) {
				Method[method] = method;
			});

			// Automatically Join
			$("#name")[0].value = random.randint(1,100);
			$("#secret")[0].value = "mime";
			login();
		}
	});
});

File: game.py
import random


class Game:
    def __init__(self, game_setting, method):
        self.game_setting = game_setting
        self.method = method

        self.started = False
        self.player_num = 0  # The number of players
        self.story_finish = 0  # The number of players that watched story
        self.captain = None  # The player that chose team

        self.users = []  # Player in the game
        self.spectating = []  # Player not in the game

        self.team = []  # Player chosen to execute mission

        self.tokens = []  # Tokens Record, 1 is good, -1 is evil
        self.token_need = []  # The number of players that tokens need
        self.approve = []  # Who aprrove the team
        self.reject = []  # Who reject the team

        self.success = []  # Who successfully execute mission
        self.fail = []  # Who failed to execute misson

        self.record = []  # Record of players choice

        self.role_map = None  # Player_id to role

        self.chosing = False  # Is captain chosing
        self.confirming = False  # Is players comfirming team

        self.round = 0
        self.failed = 0  # The number of round that reject greater than approve

    def set_tokens(self):
        self.token_need = self.game_setting["token_need"][str(len(
            self.users))]

    def get_teams(self, players):
        return [[players[i]["id"], players[i]["name"]] for i in self.users]

    def set_characters(self, connected):
        chars = self.game_setting["character_set"][str(len(self.users))]
        suf_chars = chars.copy()
        random.shuffle(suf_chars)

        self.role_map = dict(zip(self.users, suf_chars))

        response = {
            "method": self.method.START,
            "players": self.get_teams(connected),
            "token_need": self.token_need,
            "has_percival": ("PERCIVAL" in chars),
            "characters": chars}
        return response

    def has_finish_story(self):
        return len(self.users) == self.story_finish

    def chose_captain(self):
        # Clean up relative list
        self.team.clear()
        self.approve.clear()
        self.reject.clear()
        self.success.clear()
        self.fail.clear()

        if self.captain is None:
            self.captain = random.choice(self.users)
        else:
            cap_index = self.users.index(self.captain) + 1
            if cap_index == len(self.users):
                cap_index = 0

            self.captain = self.users[cap_index]

    def notify_captain(self):
        self.chosing = True
        self.confirming = False

        return {
            "method": self.method.CHOSECAPTAIN,
            "chosing": self.chosing,
            "confirming": self.confirming,
            "failed": self.failed,
            "round": self.round,
            "captain": self.captain}

    def chose_teamate(self, user_id, add_or_remove):
        if (add_or_remove and len(self.team) <
                self.token_need["numbers"][self.round]):
            if user_id not in self.team:
                self.team.append(user_id)
        else:
            if user_id in self.team:
                self.team.remove(user_id)

        response = {"method": self.method.CHOSENTEAMATE,
                    "teamates": self.team}

        return response

    def confirmTeam(self):
        self.chosing = False
        self.confirming = True

        return {
            "method": self.method.ASKAPPROVAL,
            "chosing": self.chosing,
            "confirming": self.confirming,
            "teamates": self.team}

    def vote_approve(self, _id):
        if _id not in self.approve:
            self.approve.append(_id)
        if _id in self.reject:
            self.reject.remove(_id)

        return {"method": self.method.VOTECONFIRM,
                "voter": _id}

    def vote_reject(self, _id):
        if _id not in self.reject:
            self.reject.append(_id)
        if _id in self.approve:
            self.approve.remove(_id)

        return {"method": self.method.VOTECONFIRM, "voter": _id}

    def all_voted(self):
        if len(self.approve) + len(self.reject) == len(self.users):
            return len(self.approve) > len(self.reject)
        return None

    def reset_votes(self):
        self.approve.clear()
        self.reject.clear()

    def mission_success(self, _id):
        if _id not in self.success:
            self.success.append(_id)
        if _id in self.fail:
            self.fail.remove(_id)

        return {"method": self.method.MISSIONCONFIRM, "voter": _id}

    def mission_fail(self, _id):
        if _id not in self.fail:
            self.fail.append(_id)
        if _id in self.success:
            self.success.remove(_id)

        return {"method": self.method.MISSIONCONFIRM, "voter": _id}

    def all_executed(self):
        if len(self.success) + len(self.fail) == len(self.team):
            return True
        return None

    def check_mission(self):
        record = {"success": len(self.success), "fail": len(self.fail),
                  "team": self.team.copy()}

        if self.token_need["two_fail"] == self.round and len(self.fail) <= 1:
            self.tokens.append(1)
            record["good_evil"] = "good"
        elif len(self.fail) > 0:
            self.tokens.append(-1)
            record["good_evil"] = "evil"
        else:
            self.tokens.append(1)
            record["good_evil"] = "good"

        self.record.append(record)

        self.round += 1
        self.chose_captain()

        if self.tokens.count(1) == 3:
            evils_role_map = {}

            for _id, role in self.role_map.items():
                if role in ["MORDRED", "MORGANA", "ASSASSIN", "EVIL"]:
                    evils_role_map[_id] = role
            return {"method": self.method.KILLMERLIN,
                    "evils_role_map": evils_role_map}
        elif self.tokens.count(-1) == 3:
            return self.evil_win()

    def evil_win(self, add_token=True):
        return {"method": self.method.END, "tokens": self.tokens,
                "add_token": add_token, "failed": self.failed,
                "role_map": self.role_map}

    # def good_win(self):
    #     return {"method": self.method.END, "tokens": self.tokens,
    #             "failed": self.failed, "role_map": self.role_map,
    #             "win": "good"}

    def disconnect(self, _id):
        try:
            self.users.remove(_id)
        except ValueError:
            pass
